Refactor engagement sync

pondermatic opened the pull request:
## Description
+ Moved and refactored `LLMS_User_Certificate::get_template_version()` and `LLMS_User_Certificate::sync()`  to `LLMS_Abstract_User_Engagement`.
+ Escaped single quotes in translated strings before using them in JavaScript strings.
+ Fixed typos and minor code cleanup.

Fixes #1931.

## How has this been tested?
I made sure that syncing earned certificates with a changed template worked before and after the refactoring.

I also tested syncing earned achievements by manually calling `(new LLMS_User_Achievement( 1891 ))->sync();`.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [ ] My code follows the LifterLMS Coding & Documentation Standards. UNABLE to run PHP Codesniffer at the moment.<!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

pondermatic committed:
Fix typos.

pondermatic committed:
Fix typos and minor cleanup.

pondermatic committed:
Escape single quotes in translated strings before using them in JavaScript strings.

pondermatic committed:
Refactor get_template_version() and sync() from LLMS_User_Certificate to LLMS_Abstract_User_Engagement.

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -97,6 +97,65 @@ public function get_related_post_id() {
+
+	}
+
+	/**
+	 * Retrieves the earned engagement's template version.
+	 *
+	 * Since LifterLMS 6.0.0, earned engagements are created using the block editor.
+	 *
+	 * Earned engagements created in the classic editor will use template version 1 while any earned engagements
+	 * created in the block editor use template version 2. Therefore an earned engagement that has content
+	 * and no blocks will use template version 1 and any empty earned engagements or those containing blocks
+	 * will use template version 2.
+	 *
+	 * @since [version]
+	 *
+	 * @return integer
+	 */
+	public function get_template_version() {
+
+		$version = empty( $this->get( 'content', true ) ) || has_blocks( $this->get( 'id' ) ) ? 2 : 1;
+
+		/**
+		 * Filters an earned engagement's template version.
+		 *
+		 * @since [version]
+		 *
+		 * @param int $version The template version.
+		 */
+		return apply_filters( "llms_{$this->model_post_type}_template_version", $version, $this );
+
+	}
```
I don't believe this belongs here since achievements cannot be edited in the block editor so this applied only to certificates

pondermatic replied:
`get_template_version()` is called by `sync()`, which was moved from the `LLMS_User_Certificate` class to `LLMS_Abstract_User_Engagement`. Instead of using this "generic" version, would you prefer to have a version in `LLMS_Abstract_User_Engagement` that always returns 1 and have an overriding version in `LLMS_User_Certificate`?

thomasplevy replied:
`get_template_version()` shouldn't be in the abstract `sync()` method because the metadata being synced on template version 2 is specific to certificates (they don't exist for achievements)

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -159,4 +218,83 @@ public function is_awarded() {
+			delete_post_thumbnail( $this->get( 'post' ) );
+		}
+
+		$props = array(
+			'content',
+		);
```
Duplicated below via line 282

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -159,4 +218,83 @@ public function is_awarded() {
+			'content',
+		);
+
+		// If using the block editor, also sync all layout properties.
+		if ( 2 === $template->get_template_version() ) {
+			$props = array_merge(
+				$props,
+				array(
+					'background',
+					'height',
+					'margins',
+					'orientation',
+					'size',
+					'unit',
+					'width',
+				)
+			);
+		}
+
+		foreach ( $props as $prop ) {
+			$raw = 'content' === $prop;
+			$this->set( $prop, $template->get( $prop, $raw ) );
+		}
```
All of these properties only apply to certificates

I think this method should be overloaded for certificates, call the parent method, and then run this bit of code to sync this additional meta.

pondermatic replied:
Should the overriding method also call the `llms_{$this->model_post_type}_synchronized` action hook?

pondermatic replied:
Wouldn't that mean that the base method sets the content and then the overriding method sets it again (with merged data). That would be two database writes instead of one.

thomasplevy replied:
mmm i didn't consider that... Ideally we'd want that called last.

Alright what do you think about a stub function in the abstract, something like `sync_meta()` which is responsible for performing additional metadata sync. The stub does nothing but extending classes (like certificates in this case) can override it. We can pass the existing instantiated `$template` to that class so it's available without repeating the lookup.

How's that sound?

pondermatic replied:
I like the sound of the stub method.

thomasplevy commented on includes/admin/post-types/meta-boxes/class.llms.meta.box.award.certificate.sync.php:
```diff
@@ -105,7 +105,8 @@ private function sync_action() {
 			'lifterlms'
 		);
-		$on_click   = "return confirm('${sync_alert}')";
+		$sync_alert = str_replace( "'", "\'", $sync_alert);
```
Using [esc_js()](https://developer.wordpress.org/reference/functions/esc_js/) I believe is more thorough for escaping inline JS

pondermatic replied:
`esc_js()` is too thorough. It changes `\n` to `n`.

pondermatic replied:
I tried adding various numbers of backslashes to `\n` and then running the string through `esc_js()`. It wants `\n` to be output as `\\n`, but in the Brave web browser, `\\n` shows up as a literal `\n` in the pop-up dialog.

pondermatic replied:
`esc_attr__()` may be a better solution to escaping quotes.

UPDATE: nope.

thomasplevy commented on includes/admin/post-types/meta-boxes/class.llms.meta.box.certificate.template.sync.php:
```diff
@@ -122,7 +122,8 @@ private function sync_action() {
 			$awarded_certificate_label
 		);
-		$on_click   = "return confirm('${sync_alert}')";
+		$sync_alert = str_replace( "'", "\'", $sync_alert);
```
Using [esc_js()](https://developer.wordpress.org/reference/functions/esc_js/) I believe is more thorough for escaping inline JS

thomasplevy reviewed (changes requested):
Additionally, this doesn't add any UI (metaboxes) to enable syncing of achievements. I'd like to see the existing certificate syncing UI and processing extended (and the related classes renamed to something generic where possible) so that this functionality exists for both certificates and achievements

pondermatic committed:
Fix code spaces.

pondermatic committed:
Continue refactoring sync templates to earned engagements.
- Refactor part of LLMS_Abstract_User_Engagement::sync() into LLMS_User_Certificate::sync_meta() and LLMS_Abstract_User_Engagement::sync_meta().
- Move LLMS_Abstract_User_Engagement::get_template_version() back to LLMS_User_Certificate.

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -297,4 +246,15 @@ public function sync( $context = 'update' ) {
 	}
 
+	/**
+	 * This is a stub that allows extended classes to sync block editor layout properties.
```
```suggestion
	 * This is a stub that allows extending classes to sync additional data from the template during a sync operation.
```

pondermatic committed:
Merge branch 'dev-600' into 1931-refactor-engagement-sync

thomasplevy reviewed (commented):
If `esc_js()` doesn't work let's move on. I thought it would be more thorough. WHatever :+1:

pondermatic committed:
Add handling of WP_Error objects to LLMS_Admin_Metabox.

pondermatic committed:
Refactored meta box user engagement sync classes.
- Refactored the LLMS_Meta_Box_Award_Certificate_Sync and LLMS_Meta_Box_Certificate_Template_Sync classes into LLMS_Abstract_Meta_Box_User_Engagement_Sync.
- Added the LLMS_Meta_Box_Certificate_Sync and LLMS_Meta_Box_Achievement_Sync classes.

pondermatic committed:
Add achievement sync tests.
Tweek user engagement sync model classes.

pondermatic committed:
Include abstracts for user engagements meta box, controller, processor.

pondermatic committed:
Increase the number of files that are autoloaded instead of manually loaded.
The reduces loading time (on systems with small or no opcache) and the time it takes to determine the best place to manually load a class file.

pondermatic committed:
Refactor user engagement controller classes to allow achievements to be synced with their template.

pondermatic committed:
Format translator comments in the WordPress standard.

pondermatic committed:
Refactor user engagement processor classes to allow achievements to be synced with their template.

pondermatic committed:
Refactor how user engagement strings are formatted.
Add LLMS_Trait_Case, LLMS_Trait_User_Engagement_Type, LLMS_Interface_Case, and LLMS_Interface_User_Engagement_Type.
Provide translators with engagement types that have the first letter in uppercase for use at the beginning of a sentence.

pondermatic committed:
Merge branch 'dev-600' into 1931-refactor-engagement-sync

# Conflicts:
#	includes/class-llms-loader.php

pondermatic committed:
Remove manual loading of LLMS_Block_Templates class.
The autoloader can handle it.

pondermatic commented:
Why did I change the LifterLMS autoloader you ask? While working on generating engagement processor errors, I kept having trouble with various undefined classes either while running the code or in unit testing. I've had multiple fights with the autoloader in the past and this time it was the last straw. These changes won't fix all the autoloading problems. I only fixed enough issues to get me back on the refactoring issue.
> I've always tried to go the **extra mile** with every job I work. It turns out my **Uber** customers don't appreciate it.

pondermatic commented:
@gocodebox/engineering, this PR is ready for review.

thomasplevy commented on includes/abstracts/abstract.llms.admin.metabox.php:
```diff
@@ -303,6 +304,9 @@ public function output_errors() {
 		}
 
 		foreach ( $errors as $error ) {
+			if ( is_object( $error ) && 'WP_Error' === get_class( $error ) ) {
```
```suggestion
			if ( is_wp_error( $error ) ) {
```

Prefer WP Core methods when available

thomasplevy commented on includes/class-llms-loader.php:
```diff
@@ -18,6 +18,65 @@
+		// Missing "_Abstract_" from class name.
+		'llms_admin_metabox'            => 'includes/abstracts/abstract.llms.admin.metabox.php',
+		'llms_admin_table'              => 'includes/abstracts/abstract.llms.admin.table.php',
+		'llms_analytics_widget'         => 'includes/abstracts/abstract.llms.analytics.widget.php',
```
We should be able to remove this line now:

https://github.com/gocodebox/lifterlms/blob/f1dcd79d813e1996b819f9f044ccb0f1ac67a8db/includes/admin/reporting/widgets/class.llms.analytics.widget.ajax.php#L57

pondermatic replied:
@thomasplevy,
There are dozens of include statements that can be removed with this change to the autoloader. I wasn't sure if you wanted me to go down that far into the rabbit hole, or wait until later?

thomasplevy replied:
I didn't want you to go into this rabbit hole *at all* but since we're already here and I've noticed these things then yes, we should remove them. Perhaps as a separate PR so we don't pollute this one further. Or open this as a new issue to track and clean up in the future.

pondermatic replied:
I have removed loading of all class files that don't instantiate their class in favor of autoloading.

thomasplevy commented on includes/class-llms-loader.php:
```diff
@@ -18,6 +18,65 @@
+		'llms_database_query'           => 'includes/abstracts/abstract.llms.database.query.php',
+		'llms_payment_gateway'          => 'includes/abstracts/abstract.llms.payment.gateway.php',
+		'llms_post_model'               => 'includes/abstracts/abstract.llms.post.model.php',
+		'llms_shortcode_course_element' => 'includes/abstracts/abstract.llms.shortcode.course.element.php',
```
https://github.com/gocodebox/lifterlms/blob/858811569e522b612e89f8d9ed29f5b36cb2b8aa/includes/shortcodes/class.llms.shortcodes.php#L62

thomasplevy commented on includes/class-llms-loader.php:
```diff
@@ -18,6 +18,65 @@
+		'llms_payment_gateway'          => 'includes/abstracts/abstract.llms.payment.gateway.php',
+		'llms_post_model'               => 'includes/abstracts/abstract.llms.post.model.php',
+		'llms_shortcode_course_element' => 'includes/abstracts/abstract.llms.shortcode.course.element.php',
+		'llms_shortcode'                => 'includes/abstracts/abstract.llms.shortcode.php',
```
https://github.com/gocodebox/lifterlms/blob/858811569e522b612e89f8d9ed29f5b36cb2b8aa/includes/shortcodes/class.llms.shortcodes.php#L61

thomasplevy commented on includes/class-llms-loader.php:
```diff
@@ -18,6 +18,65 @@
+		'llms_post_model'               => 'includes/abstracts/abstract.llms.post.model.php',
+		'llms_shortcode_course_element' => 'includes/abstracts/abstract.llms.shortcode.course.element.php',
+		'llms_shortcode'                => 'includes/abstracts/abstract.llms.shortcode.php',
+		'llms_update'                   => 'includes/abstracts/abstract.llms.update.php',
```
Self note: I believe this is a completely unused class that's been replaced and can be deprecated and later removed

pondermatic replied:
I created an issue for the unused LLMS_Update class.

thomasplevy commented on includes/traits/llms-trait-case.php:
```diff
@@ -0,0 +1,53 @@
+<?php
```
okay I see the purpose here and I like the utility of this but I'm also thinking that this could be a useful function outside classes... Since the usage of a trait necessitates using an interface to define the constants it also feels a bit cumbersome to mix into a class: I mean you can't just `use LLMS_Trait_Case` you also have to implement the interface...

I'm therefore considering that my initial thinking, that this is just a function, could be modified and merged with the reasons you've noted for using a trait (class) -- using constants as opposed to strings is superior and keeping it within a class prevents global scope constant "pollution"

So... could this be a class with a single (public) static function: `public static change_case()`.

All the constants can be defined in the one class/file without the need of the interface and it would also expose the functionality.

thomasplevy commented on includes/admin/class.llms.admin.menus.php:
```diff
@@ -5,7 +5,7 @@
  *
  * @since 1.0.0
- * @version 5.3.1
+ * @version [version]
```
The changes in this file break the status tools page since the tools are no longer loaded.

I think these changes should be reverted

In the `status_page_includes()` function there are two referenced classes that can be removed though since they're now autoladed:

require_once LLMS_PLUGIN_DIR . 'includes/abstracts/llms-abstract-admin-tool.php';

and

require_once LLMS_PLUGIN_DIR . 'includes/abstracts/llms-abstract-admin-tool.php';

pondermatic replied:
Thank you for catching this. It looks like I had intended to autoload the tools, but since those class files instantiate themselves, I decided **not** to address that kind of issue in this round of autoloader changes.

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+			'sync_one'  => "sync_awarded_$this->engagement_type",
+			'sync_many' => "sync_awarded_{$this->engagement_type}s",
+		);
+		if ( ! isset( $_GET['action'] ) ) {
```
```suggestion
		$action = llms_filter_input( INPUT_GET, 'action' );
		if ( ! $action ) {
```

pondermatic replied:
Good suggestion.

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+				"llms-sync-awarded-{$this->engagement_type}s-missing-action",
+				__( 'Sorry, you have not provided any actions.', 'lifterlms' )
+			);
+		} elseif ( ! in_array( $_GET['action'], $actions ) ) {
```
```suggestion
		} elseif ( ! in_array( $action, $actions, true ) ) {
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+
+		$engagement_id = llms_filter_input( INPUT_GET, 'post', FILTER_SANITIZE_NUMBER_INT );
+
+		if ( $_GET['action'] === $actions['sync_one'] ) {
```
```suggestion
		if ( $action === $actions['sync_one'] ) {
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+			} else {
+				$result = $this->sync_awarded_engagements( $engagement_id );
+			}
+		} else {
+			$result = null;
```
```suggestion
```

We can't get here because $action must be in the $actions array in order to proceed past the initial validation

pondermatic replied:
@thomasplevy,
True, but some code sniffing tools (I'm looking at you PhpStorm!) can't figure that out and report that `$result` may be undefined. Another option is to set `$result = null` before that `if` block.

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+			$result = null;
+		}
+
+		if ( is_object( $result ) && 'WP_Error' === get_class( $result ) ) {
```
```suggestion
		if ( is_wp_error( $result ) ) {
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+			} else {
+				$result = $this->sync_awarded_engagement( $engagement_id );
+			}
+		} elseif ( $_GET['action'] === $actions['sync_many'] ) {
```
```suggestion
		} elseif ( $action === $actions['sync_many'] ) {
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+		}
+
+		$engagement_id = llms_filter_input( INPUT_GET, 'post', FILTER_SANITIZE_NUMBER_INT );
+
```
I'd like to see the two `empty( $engagement_id )` checks / errors combined into a single error here. They're basically both the same except for the post type name. I would suggest moving the generation of the error message into a private helper function, something like `get_missing_post_id_error_message( $action )` or something like that... Seems like a lot of redundant code to me is all.

pondermatic replied:
@thomasplevy,
With the replacement of `LLMS_Trait_User_Engagement_Type::get_engagement_type_name()` with `LLMS_Abstract_Controller_User_Engagements::get_text()`, generating the error message is much more compact and easier to follow. There's still a fair bit of indentation going on.
```php
if ( $action === $actions['sync_one'] ) {
    if ( empty( $engagement_id ) ) {
        $result = new WP_Error(
            "llms-sync-missing-awarded-$this->engagement_type-id",
            $this->get_text( self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID )
        );
    } else {
        $result = $this->sync_awarded_engagement( $engagement_id );
    }
} elseif ( $action === $actions['sync_many'] ) {
    if ( empty( $engagement_id ) ) {
        $result = new WP_Error(
            "llms-sync-missing-{$this->engagement_type}-template-id",
            $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID )
        );
    } else {
        $result = $this->sync_awarded_engagements( $engagement_id );
    }
} else {
    $result = null;
}
```

How about this? Is it any better?
```php
$is_syncing_one = $action === $actions['sync_one'];
if ( empty( $engagement_id ) ) {
    if ( $is_syncing_one ) {
        $code    = "llms-sync-missing-awarded-$this->engagement_type-id";
        $message = $this->get_text( self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID );
    } else {
        $code    = "llms-sync-missing-{$this->engagement_type}-template-id";
        $message = $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID );
    }
    $result = new WP_Error( $code, $message );
} else {
    if ( $is_syncing_one ) {
        $result = $this->sync_awarded_engagement( $engagement_id );
    } else {
        $result = $this->sync_awarded_engagements( $engagement_id );
    }
}
```

pondermatic replied:
I committed the last option.

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+				$result = new WP_Error(
+					"llms-sync-missing-awarded-$this->engagement_type-id",
+					sprintf(
+						/* translators: 1: singular lowercase awarded engagement type, 2: singular uppercase first letter awarded engagement type */
+						_x( 'Sorry, you need to provide a valid %1$s ID.', 'awarded', 'lifterlms' ),
+						$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::AWARDED ),
+						$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::AWARDED )
+					)
```
This string doesn't contain a `%2$s` in it

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+				$result = new WP_Error(
+					"llms-sync-missing-{$this->engagement_type}-template-id",
+					sprintf(
+						/* translators: 1: singular lowercase engagement template type, 2: singular uppercase first letter engagement template type */
+						_x( 'Sorry, you need to provide a valid %1$s ID.', 'template', 'lifterlms' ),
+						$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::TEMPLATE ),
+						$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::TEMPLATE )
+					)
```
string doesn't contain `%2$s`

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+		if ( ! current_user_can( 'edit_post', $engagement_id ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-$this->engagement_type-insufficient-permissions",
+				sprintf(
+					/* translators: 1: singular lowercase awarded engagement type, 2: awarded engagement ID, 3: singular uppercase first letter awarded engagement type */
+					__( 'Sorry, you are not allowed to edit the awarded %1$s #%2$d.', 'lifterlms' ),
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::AWARDED ),
+					$engagement_id,
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::AWARDED )
+				),
```
The number of placeholders is messed up here

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+
+		$sync = $this->get_awarded_engagement( $engagement_id )->sync();
+		if ( ! $sync ) {
+			return new WP_Error(
+				"llms-sync-awarded-$this->engagement_type-invalid-template",
+				sprintf(
+					/* translators: 1: singular lowercase awarded engagement type, 2: awarded engagement ID, 3: singular lowercase engagement template type, 4: singular uppercase first letter awarded engagement type, 5: singular uppercase first letter engagement template type */
+					__( 'Sorry, the %1$s #%2$d does not have a valid %3$s.', 'lifterlms' ),
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::AWARDED ),
+					$engagement_id,
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::TEMPLATE ),
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::AWARDED ),
+					$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::TEMPLATE )
+				),
+				compact( 'engagement_id' )
+			);
```
placeholder count is messed up

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,234 @@
+		if ( ! current_user_can( get_post_type_object( "llms_my_$this->engagement_type" )->cap->edit_posts ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-insufficient-permissions",
+				sprintf(
```
mismatched placeholders

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -28,17 +40,17 @@ abstract class LLMS_Abstract_User_Engagement extends LLMS_Post_Model {
+		 * @param LLMS_Abstract_User_Engagement $User_Engagement Achievement or certificate class object.
 		 */
-		do_action( "llms_before_delete_{$this->model_post_type}", $this );
+		do_action( "llms_before_delete_$this->model_post_type", $this );
```
```suggestion
		do_action( "llms_before_delete_{$this->model_post_type}", $this );
```

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#interpolation-for-naming-dynamic-hooks

In general this PR has a lot of variable interpolation within strings that *does not* use curly braces. I think it's easier to read a variable in curly braces... Perhaps it's not hte preferred general method but this is the convention we're following, please use curly braces where possible.

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -28,17 +40,17 @@ abstract class LLMS_Abstract_User_Engagement extends LLMS_Post_Model {
 		 *
-		 * They dynamic portion of this hook, `{$this->model_post_type}`, refers to the engagement type,
-		 * either "achievement" or "certificate".
+		 * They dynamic portion of the hook name, {@see LLMS_Abstract_User_Engagement::$model_post_type `$this->model_post_type`},
+		 * refers to the engagement type, either "achievement" or "certificate".
```
I know this is better for your IDE but it's going to look weird when rendered on the codex site with the reference like that...

pondermatic replied:
You're right. After testing, I found that it looks good if the curly braces are wrapped in backticks, and the description part is left out. But the link doesn't work, probably because it is to a protected property.

I will remove the `@see` tags from hook DocBlocks.

thomasplevy commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -54,17 +66,17 @@ public function delete() {
 		 */
-		do_action( "llms_delete_{$this->model_post_type}", $this );
-
+		do_action( "llms_delete_$this->model_post_type", $this );
```
```suggestion
		do_action( "llms_delete_{$this->model_post_type}", $this );
```

thomasplevy commented on includes/abstracts/llms-abstract-processor-user-engagement-sync.php:
```diff
@@ -0,0 +1,304 @@
+			sprintf(
+				/* translators: 1: plural uppercase first letter awarded engagement type, 2: opening anchor tag that links to the engagement template, 3: engagement template name, 4: engagement template ID, 5: closing anchor tag, 6: plural lowercase awarded engagement type */
+				__( '%1$s sync completed for the template %2$s%3$s (#%4$d)%5$s.', 'lifterlms' ),
+				$this->get_engagement_type_name( self::PLURAL, self::CASE_UPPER_FIRST, self::AWARDED ),
```
The completion notice for both certificates & achievements when bulk syncing from template -> awarded versions resulted in "Unknown Engagement Type sync completed...". I'm not sure where the problem lies exactly but I saw it on both of them

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+			$plural_or_singular                       = self::SINGULAR;
+		}
+
+		$this->title   = sprintf(
```
```suggestion
		$this->title = sprintf(
```

CS

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+	 *
+	 * @return string
+	 */
+	private function sync_action() {
```
Can we split this into another method, one for the first condition and another for the second. I guess a third too for the empty condition handled in `output()`.

This is just a really big method

pondermatic replied:
I agree. Done.

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+
+			$plural_or_singular = ( $awarded_number > 1 ) ? self::PLURAL : self::SINGULAR;
+			$sync_alert         = sprintf(
+				/* translators: 1: number of awarded engagements, 2: plural or singular lowercase awarded engagement type, 3: singular lowercase engagement template type, 4: plural or singular uppercase first letter awarded engagement type, 5: singular uppercase first letter engagement template type */
```
wrong number of placeholders

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+				$this->get_engagement_type_name( $plural_or_singular, self::CASE_UPPER_FIRST, self::AWARDED ),
+				$this->get_engagement_type_name( self::SINGULAR, self::CASE_UPPER_FIRST, self::TEMPLATE )
+			);
+			$sync_description   = sprintf(
```
wrong number of placeholders

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+				return '';
+			}
+
+			$sync_alert       = sprintf(
```
wrong number of placeholders

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+			$sync_description = sprintf(
+				/* translators: 1: singular lowercase awarded engagement type, 2: link to edit the engagement template, 3: singular lowercase engagement template type, 4: closing anchor tag, 5: singular uppercase first letter awarded engagement type, 6: singular uppercase first letter engagement template type */
+				__( 'Sync this %1$s with its %2$s%3$s%4$s.', 'lifterlms' ),
+				$this->get_engagement_type_name( self::SINGULAR, self::CASE_LOWER, self::AWARDED ),
```
wrong number of placeholders

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+		$base_url = remove_query_arg( 'action' ); // Current URL without 'action' arg.
+		$sync_url = add_query_arg(
+			'action',
+			"sync_awarded_$this->engagement_type" . ( $this->is_syncing_all_awarded_engagements ? 's' : '' ),
```
```suggestion
			"sync_awarded_{$this->engagement_type}" . ( $this->is_syncing_all_awarded_engagements ? 's' : '' ),
```

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,263 @@
+			"sync_awarded_$this->engagement_type" . ( $this->is_syncing_all_awarded_engagements ? 's' : '' ),
+			wp_nonce_url(
+				$base_url,
+				"llms-$this->engagement_type-sync-actions",
```
```suggestion
				"llms-{$this->engagement_type}-sync-actions",
```

pondermatic committed:
Refactor LLMS_Trait_Case to LLMS_Case.

thomasplevy commented on includes/admin/post-types/meta-boxes/class.llms.meta.box.achievement.sync.php:
```diff
@@ -0,0 +1,56 @@
+<?php
```
I didn't want to mess with this on roccos initial work but since the autoloader will now pick this up with either space or dash delimiter let's change the filename to be dashes (as we intend new files to be)

thomasplevy commented on includes/admin/post-types/meta-boxes/class.llms.meta.box.certificate.sync.php:
```diff
@@ -0,0 +1,56 @@
+<?php
```
Rename file to use dashes instead of spaces

thomasplevy commented on includes/processors/class.llms.processor.achievement.sync.php:
```diff
@@ -0,0 +1,44 @@
+<?php
```
filename should use dashes instead of dots

pondermatic replied:
The processor class files instantiate the class and are not currently autoloaded. I'm saving that for later.

pondermatic replied:
I decided to modify `LLMS_Processors::load_processor()` to handle class file names with dashes.

@thomasplevy, does this change need a `.changelogs` files?

pondermatic replied:
NVM. It's a brand new file.

thomasplevy commented on includes/processors/class.llms.processor.certificate.sync.php:
```diff
@@ -0,0 +1,44 @@
+<?php
```
filename should use dashes instead of dots

thomasplevy commented on includes/class-llms-case.php:
```diff
@@ -0,0 +1,108 @@
+	 * @param string $case   One of the CASE_ constants from {@see LLMS_Case}.
+	 * @return string
+	 */
+	public static function change( $string, $case ) {
```
This looks great, only change I'd suggest would be to split the conditions into individual (private) functions: 1 for multibyte function usage and one for non mb usage. This will make testing easier while also slimming down the method's length.

pondermatic committed:
Split LLMS_Case::change() into change_with_multibyte() and change_without_multibyte().

pondermatic commented:
@thomasplevy,

### The short version

Translated phrases should **not** be inserted in a translated sentence. From the WordPress [Best Practices for writing strings](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#best-practices-for-writing-strings)

> Use entire sentences – in most languages word order is different than that in English.

I'm going to added functions in the final classes that return a translated full sentence, without inserting phrases.

I'm not sure `LLMS_Case` has much value. Maybe for logs, but using it with translations may not be a good idea after all.

### The long version

The extra translator `sprintf()` arguments were to allow translators to begin the sentence with the translated phrase. For example, in some languages

> Sync awarded certificates with the certificate template.

might need to be reordered to

> Awarded certificates shall be synced with the certificate template.

While writing unit tests for the multibyte versions of [`ucfirst()`](https://www.php.net/manual/en/function.ucfirst.php), I discovered that changing the case of "Awarded Certificate", from `LLMS_Post_Types`, to "awarded certificate" and "Awarded certificate" **may** not be a good idea. For example, in Greek, ["Awarded" = "Απονεμήθηκε", but "awarded" = "απονεμηθεί"](https://translate.google.com/?sl=en&tl=el&text=Awarded%20Certificate%0Aawarded%20certificate%0AAwarded%20certificate%0A&op=translate). Confusingly, changing the case of "Απονεμήθηκε Πιστοποιητικό" [seems to work](https://translate.google.com/?sl=el&tl=en&text=%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CE%A0%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%B1%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C&op=translate)?

pondermatic committed:
Move creation of user text from abstract engagement classes to final engagement classes.
Injecting translated phrases into a string that is being translated creates problems when translating to languages that may need to move the phrase to the beginning, or use a different phrase depending on the context of where it is in the sentence.

pondermatic committed:
Use is_wp_error().

pondermatic committed:
Revert deprecation of LLMS_Admin_Menus::status_page_includes().
It can't go away until the admin tools files don't instantiate themselves.

pondermatic committed:
Wrap properties in curly braces when used in double-quoted strings.

pondermatic committed:
Dynamic hooks should be named using interpolation rather than concatenation for readability and discoverability purposes.

pondermatic committed:
Fix typos.

pondermatic committed:
Clean up based on code review suggestions.

pondermatic committed:
Refactor LLMS_Trait_User_Engagement_Type::get_awarded_engagement() and LLMS_Abstract_Meta_Box_User_Engagement_Sync::get_engagement_template() into LLMS_Trait_User_Engagement_Type::get_user_engagement().

pondermatic committed:
Add ability to autoload analytic widgets.

pondermatic committed:
Add ability to autoload reporting tables.

pondermatic committed:
Cleanup DocBlocks.

pondermatic committed:
Add ability to autoload notification views.

pondermatic committed:
Cleanup DocBlocks.

pondermatic committed:
Add ability to autoload privacy classes.

pondermatic committed:
Warn developers about the deprecated LLMS_Achievement class.

pondermatic committed:
Removed loading of deprecated class files.

pondermatic committed:
Fix typos in get_settings_group_achievements().

pondermatic committed:
Return early from autoload() if not a LifterLMS core class.

pondermatic committed:
Add LLMS_Loader::$class_paths property to clean up autoload() code.

pondermatic committed:
Removed loading of deprecated class files.

pondermatic committed:
Warn developers about the deprecated LLMS_Certificate class.

pondermatic committed:
Restore ability to load 3rd party views.

pondermatic committed:
Add ability to autoload email classes.

pondermatic committed:
Add ability to autoload meta box field classes.

pondermatic committed:
Removed loading of class files that don't instantiate their class in favor of autoloading.

pondermatic committed:
Tweak deprecation notice of LLMS_Achievement and LLMS_Certificate for PHPUnit compatibility.

pondermatic committed:
Tweak deprecation notices.
_deprecated_function() prefixes the replacement message with "Use " and postfixes it with " instead.";

pondermatic committed:
Replace use of the deprecated `LLMS_Certificate` class

pondermatic committed:
Rename new class files according to the LifterLMS standard.

thomasplevy commented on includes/admin/class.llms.admin.post-types.php:
```diff
@@ -70,15 +70,6 @@ public function use_block_editor_for_post( $use_block_editor, $post ) {
 
 	}
 
-	/**
-	 * Initializes core for metaboxes
-	 *
-	 * @return void
-	 */
-	public function meta_metabox_init() {
-		include_once 'llms.class.admin.metabox.php';
-	}
-
```
This should be deprecated in favor of removed outright to bar against the fatal error from 3rd parties... Though unlikely as it is that this is being used externally.

thomasplevy commented on includes/class.llms.achievement.php:
```diff
@@ -124,8 +124,16 @@ public function __call( $name, $args ) {
+	 */
 	public function __construct() {
 
+		_deprecated_function( __METHOD__, '[version]' );
```
I think the idea here is that the whole class is deprecated so adding a deprecation notice to the constructor alerts developers. In which case I think we should add a message and instead of outputting that the constructor is deprecated, we should output that the class is deprecated
```suggestion
		_deprecated_function( __CLASS__, '[version]' );
```

pondermatic replied:
Interesting! I didn't try just the class name. Initially I did this.
```php
// Log "The LLMS_Achievement class is deprecated since version [version] with no alternative available.".
_deprecated_function( 'The ' . __CLASS__ . ' class', '[version]' );
```
However, when I added `@expectedDeprecated LLMS_Certificate::__construct` to [`LLMS_Test_LLMS_User_Certificate::test_get_merge_data_deprecated_hook()`](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/tests/phpunit/unit-tests/models/class-llms-test-model-llms-user-certificate.php#L648) I got errors.

Providing just the class name seem to work very well.

thomasplevy commented on includes/class.llms.certificate.php:
```diff
@@ -132,10 +132,15 @@ public function __call( $name, $args ) {
 	 */
 	public function __construct() {
 
+		_deprecated_function( __METHOD__, '[version]' );
```
```suggestion
		_deprecated_function( __CLASS__, '[version]' );
```

thomasplevy commented on includes/class.llms.certificates.php:
```diff
@@ -84,18 +84,16 @@ private function __construct() {
 	public function init() {
 
-		include_once 'class.llms.certificate.php';
-		$this->certs['LLMS_Certificate_User'] = isset( $this->certs['LLMS_Certificate_User'] ) ? $this->certs['LLMS_Certificate_User'] : include_once 'certificates/class.llms.certificate.user.php';
```
With this change `llms()->certificates()->certs['LLMS_Certificate_User']` will result in an error. For now this line should be left intact until `LLMS_Certificate_User` is fully removed.

pondermatic replied:
@thomasplevy,
Loading `class.llms.certificate.php` in line 96 is handled by the autoloader now.

Loading `certificates/class.llms.certificate.user.php` in line 97 will create a deprecation warning from [`LLMS_Certificate::_construct()`](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/class.llms.certificate.php#L142). Also, the [autoloader](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/class-llms-loader.php#L90) handles that class file.

thomasplevy replied:
> Loading `class.llms.certificate.php` in line 96 is handled by the autoloader now.

I intended to only make a note on line 97, sorry.



> Loading `certificates/class.llms.certificate.user.php` in line 97 will create a deprecation warning from [`LLMS_Certificate::_construct()`](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/class.llms.certificate.php#L142). Also, the [autoloader](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/class-llms-loader.php#L90) handles that class file.

But if a dev accesses the property it'll create a fatal error because the the public class property isn't available.

pondermatic replied:
Now I understand why `LLMS_Certificate::__construct()` and `LLMS_Achievement::__construct()` didn't emit deprecation warnings. I will undo the `__construct()` deprecation warnings. The object will be there and will emit deprecation warnings if a 3rd party tries to call any method.

thomasplevy commented on includes/functions/llms-functions-deprecated.php:
```diff
@@ -324,5 +324,5 @@ function llms_engagement_handle_deprecated_meta_keys( $val, $obj_id, $key ) {
 	$classname = get_class( $obj );
 	$keyname   = strtolower( str_replace( 'LLMS_User_', '', $classname ) ) . '_' . $meta_key;
-	_deprecated_function( "{$classname} meta key '{$keyname}'", '[version]',  );
+	_deprecated_function( "{$classname} meta key '{$keyname}'", '[version]', $replacement_msg );
```
:+1:

thomasplevy commented on includes/shortcodes/class.llms.shortcodes.php:
```diff
@@ -34,9 +35,9 @@ public static function init() {
 			 * Filters the shortcodes to initialize
 			 *
-			 * @since unknown
+			 * @since Unknown.
```
```suggestion
			 * @since Unknown
```

Unknown as the description should have a period, Unknown as the version should not have a period

Eg:

```
@since Unknown
@since 1.0.0 Unknown.
```

thomasplevy commented on includes/shortcodes/class.llms.shortcodes.php:
```diff
@@ -68,9 +67,9 @@ public static function init() {
 			 * Filters the path of the shortcode class file
 			 *
-			 * @since unknown
+			 * @since Unknown.
```
```suggestion
			 * @since Unknown
```

thomasplevy commented on includes/shortcodes/class.llms.shortcodes.php:
```diff
@@ -98,27 +97,24 @@ public static function init() {
 				 * The dynamic portion of the hook name, `$shortcode` refers to the shortcode tag itself.
 				 *
-				 * @since unknown
+				 * @since Unknown.
```
```suggestion
				 * @since Unknown
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,252 @@
+				$message = $this->get_text( self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID );
+			}
+			$result = new WP_Error( $code, $message );
+		} else {
+			if ( $is_syncing_one ) {
+				$result = $this->sync_awarded_engagement( $engagement_id );
+			} else {
+				$result = $this->sync_awarded_engagements( $engagement_id );
+			}
+		}
```
```suggestion
		} elseif ( $is_syncing_one ) {
			$result = $this->sync_awarded_engagement( $engagement_id );
		} else {
			$result = $this->sync_awarded_engagements( $engagement_id );
		}
```

pondermatic replied:
Even better!

thomasplevy commented on includes/controllers/class.llms.controller.achievements.php:
```diff
@@ -26,48 +36,67 @@ class LLMS_Controller_Achievements {
+		switch ( $text_type ) {
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS:
+				return sprintf(
+					/* translators: 1: awarded achievement ID */
```
This assumes the translator can infer that "%1$d" is the same as "1". Best practice is to use the actual placeholder being used in favor of just the placeholder's number.
```suggestion
					/* translators: %1$d: awarded achievement ID */
```

I am not aware of any specific rule about this but if you look through WP source they seem to exclusively follow this pattern (and that's what we intend to do as well)

pondermatic replied:
@thomasplevy,
I agree that the full placeholder is helpful to translators. I was following the examples in [How to Internationalize Your Plugin](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#variables).

There are 423 examples in WordPress code that use just the numbers, but only 2 that use the full placeholder.

thomasplevy replied:
@pondermatic great, do what I'm asking anyway here <3

thomasplevy commented on includes/controllers/class.llms.controller.achievements.php:
```diff
@@ -26,48 +36,67 @@ class LLMS_Controller_Achievements {
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE:
+				return sprintf(
+					/* translators: 1: awarded achievement ID */
```
```suggestion
					/* translators: %1$d: awarded achievement ID */
```

thomasplevy commented on includes/controllers/class.llms.controller.certificates.php:
```diff
@@ -11,34 +11,80 @@
+		switch ( $text_type ) {
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS:
+				return sprintf(
+					/* translators: 1: awarded certificate ID */
```
```suggestion
					/* translators: %1$d: awarded certificate ID */
```

thomasplevy commented on includes/controllers/class.llms.controller.certificates.php:
```diff
@@ -11,34 +11,80 @@
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE:
+				return sprintf(
+					/* translators: 1: awarded certificate ID */
```
```suggestion
					/* translators: %1$d: awarded certificate ID */
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,252 @@
+		 *
+		 * @since [version]
+		 *
+		 * @see   LLMS_Processor_Certificate_Sync.
```
```suggestion
		 * @see LLMS_Processor_Certificate_Sync.
```

thomasplevy commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,252 @@
+		}
+
+		/**
+		 * Fires an action to trigger the bulk sync of awarded engagements.
+		 *
+		 * @since [version]
```
```suggestion
		 * Fires an action to trigger the bulk sync of awarded engagements.
		 *
		 * The dynamic portion of this hook, `{$this->engagement_type}`, refers to the type of awarded engagement, either "achievement" or "certificate".
		 *
		 * @since [version]
```

pondermatic committed:
Deprecate LLMS_Admin_Post_Types::meta_metabox_init() instead of removing it.

pondermatic committed:
Improve deprecation warning on the LLMS_Achievement and LLMS_Certificate classes.

pondermatic committed:
Remove period from "@since Unknown".

pondermatic committed:
Improve if logic in maybe_handle_awarded_engagement_sync_actions().

thomasplevy commented on includes/class.llms.achievements.php:
```diff
@@ -73,14 +73,12 @@ private function __construct() {
 	public function init() {
-
-		include_once 'class.llms.achievement.php';
-		$this->achievements['LLMS_Achievement_User'] = include_once 'achievements/class.llms.achievement.user.php';
```
Same as certificates class

pondermatic committed:
Merge branch 'dev-600' into 1931-refactor-engagement-sync

pondermatic committed:
Revert the `__construct()` deprecation warnings.
For now, `LLMS_Certificates::init()` needs to instantiate `LLMS_Certificate_User`, extended from `LLMS_Certificate`, in case 3rd parties try to use it. They will get a deprecation warning when calling any method.

pondermatic committed:
Use the full placeholder in translator comments.

pondermatic committed:
Merge remote-tracking branch 'origin/1931-refactor-engagement-sync' into 1931-refactor-engagement-sync

pondermatic committed:
Remove @see tag from hook DocBlocks.
Can't link to a protected property.

pondermatic committed:
Refactor sync_action() to create get_sync_action_texts().

pondermatic committed:
Add the ability to load processor class files with dashes in their file name.

pondermatic committed:
Fix DocBlock.

thomasplevy commented:
Doing manual testing on syncing functionality. It seems to all be syncing properly but I got this success message

![Screenshot_2022-02-08_14-33-38](https://user-images.githubusercontent.com/1290739/153079338-34b9b9dd-f48b-431c-a7d3-da305ef5dc3f.png)

Edit: same issue with achievements too

thomasplevy commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,309 @@
+			}
+
+			$sync_alert       = $this->get_text( self::TEXT_SYNC_ALERT_THIS_AWARDED_ENGAGEMENT );
+			$sync_description = $this->get_text( self::TEXT_SYNC_DESCRIPTION_THIS_AWARDED_ENGAGEMENT );
```
No vars are being passed resulting in the current post being used for the link in the metabox (in favor of the link to the template)

pondermatic replied:
@thomasplevy, Thank you for catching this.

thomasplevy commented on includes/llms.functions.core.php:
```diff
@@ -836,6 +837,11 @@ function llms_get_post( $post, $error = false ) {
 		}
 	}
 
+	// Engagement template post types use a different class as of LifterLMS 6.0.0.
+	if ( 'LLMS_Achievement' === $class || 'LLMS_Certificate' === $class ) {
+		$class = 'LLMS_User_' . substr( $class, 5 );
+	}
+
```
Hmmm I'm not sure about this... It does make sense to make this function work for achievements & certificates, however most of 6.0 code has been relying on `llms_get_certificate()` and hasn't needed an equivalent `llms_get_achievement()`... Furthermore this makes it so that we can only pass in template posts and not awarded (`_my_` versions).

Historically, the `LLMS_User_` prefix breaks the established model class pattern because both `LLMS_Achievement` and `LLMS_Certificate` classes already existed so we couldn't use the same name. I guess a developer might expect `llms_get_post()` to work this way but it's also not essential as far as our internal code base is concerned.

Are you seeing a need for this somewhere that I'm missing?

pondermatic replied:
@thomasplevy, I didn't like this change either. It feels like an ugly hack.

I'm pretty sure it was put in to help failing unit tests caused by the, now reverted, addition of the deprecated `LLMS_Achievement` and `LLMS_Certificate` constructors? I'll revert this change.

thomasplevy replied:
Marking as unresolved to help track changes that have been requested.

eri-trabiccolo commented on includes/abstracts/abstract.llms.post.model.php:
```diff
@@ -308,7 +308,7 @@ protected function after_create() {}
 	 */
 	private function create( $title = '' ) {
-		return wp_insert_post( wp_slash( apply_filters( 'llms_new_' . $this->model_post_type, $this->get_creation_args( $title ) ) ), true );
+		return wp_insert_post( wp_slash( apply_filters( "llms_new_{$this->model_post_type}", $this->get_creation_args( $title ) ) ), true );
```
Maybe we can take this occasion to add a the documentation to this filter?

thomasplevy replied:
In the interest of time let's not :+1:

thomasplevy replied:
https://github.com/gocodebox/lifterlms/issues/1982

pondermatic replied:
Tracked in issue #1982.

eri-trabiccolo commented on includes/abstracts/abstract.llms.post.model.php:
```diff
@@ -1036,9 +1036,9 @@ public function get_status_name() {
 		 *
 		 * @since 3.0.0
 		 *
```
Mind to add also the classic
`The dynamic portion of this hook...`
since we're there?

pondermatic replied:
This hook is already documented with that sentence.

eri-trabiccolo replied:
I missed it, my bad ❤️

eri-trabiccolo commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,253 @@
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS = 0;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE = 1;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENTS_INSUFFICIENT_PERMISSIONS = 2;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENTS_INVALID_NONCE = 3;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID = 4;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID = 5;
```
no summary for constants?
Also, from our documentation standards, we don't seem to leave an empty line between the `@since` tag and the `@var` tag.

thomasplevy replied:
There should be an empty line regardless of whether or not the standards indicate as much.

pondermatic replied:
I looked and didn't find summary requirement in the LifterLMS or WordPress documentation standards, but the current version in the proposed [PHP FIG PSR](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#51-summary) says:
> A Summary MUST contain an abstract of the "Structural Element" defining the purpose.

I will add summaries to the new constants.

From the [LifterLMS documentation standard](https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md#tag-alignment-and-order):
> All changelog tags, `@since`, `@version`, and `@deprecated` should be grouped together with a space before the first `@since` tag and after the last tag in the group.

eri-trabiccolo replied:
Fine for the docs standard, only the examples there are misleading. Once again my bad :)

eri-trabiccolo commented on includes/abstracts/llms-abstract-controller-user-engagements.php:
```diff
@@ -0,0 +1,253 @@
+		} elseif ( ! in_array( $action, $actions, true ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-invalid-action",
+				__( 'You\'re trying to perform an invalid action.', 'lifterlms' )
```
```suggestion
				__( "You're trying to perform an invalid action.", 'lifterlms' )
```
Thomas told me once to avoid, where possible, to use _(back)slashes_ in localized strings, to avoid translators messing with them.

eri-trabiccolo commented on includes/controllers/class.llms.controller.achievements.php:
```diff
@@ -26,48 +36,67 @@ class LLMS_Controller_Achievements {
 
-		if ( isset( $_POST['llms_delete_achievement'] ) ) {
-			$this->delete( llms_filter_input( INPUT_POST, 'achievement_id', FILTER_SANITIZE_NUMBER_INT ) );
+		switch ( $text_type ) {
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS:
+				return sprintf(
+					/* translators: %1$d: awarded achievement ID */
+					__( 'Sorry, you are not allowed to edit the awarded achievement #%1$d.', 'lifterlms' ),
+					( $variables['engagement_id'] ?? 0 )
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE:
+				return sprintf(
+					/* translators: %1$d: awarded achievement ID */
+					__( 'Sorry, the awarded achievement #%1$d does not have a valid achievement template.', 'lifterlms' ),
+					( $variables['engagement_id'] ?? 0 )
+				);
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENTS_INSUFFICIENT_PERMISSIONS:
+				return __( 'Sorry, you are not allowed to edit awarded achievements.', 'lifterlms' );
+			case self::TEXT_SYNC_AWARDED_ENGAGEMENTS_INVALID_NONCE:
+				return __( 'Sorry, you are not allowed to sync awarded achievements.', 'lifterlms' );
+			case self::TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID:
+				return __( 'Sorry, you need to provide a valid awarded achievement ID.', 'lifterlms' );
+			case self::TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID:
+				return __( 'Sorry, you need to provide a valid achievement template ID.', 'lifterlms' );
+			default:
+				return parent::get_text( $text_type );
```
Do we really need all this here, and in the certificates controller? Aren't they basically the same method with the only exception being the post type? Can't we just use then the post type labels (singular and plural) and move this method in the abstract class?

pondermatic replied:
@eri-trabiccolo, yes. The nearly identical `get_text()` methods are needed in `LLMS_Controller_Achievements` and `LLMS_Controller_Certificates` because translations should be of sentences and not of phrases. 

See https://github.com/gocodebox/lifterlms/pull/1946#issuecomment-1030073416.

I initially moved all the text to the abstract classes and injected the appropriate post type and singular or plural. It turned out to be bad for translators.

eri-trabiccolo replied:
I don't think we're in the sentence/phrase case here. Are we?
Also:
1) I think that `awarded achievement(s)` still is the post type label (plural or singular), so if we change it that string should change.
2) `It turned out to be bad for translators.`. How bad? I don't think it's worse, with your solution translators have to basically double the translations. I do understand that `awarded achievement` can be feminine and `awarded certificate` can be masculine in some languages, and some languages have feminine and masculine articles (Italian has feminine and masculine articles - both definite and indefinite), in this case is up to the translator, to me, to find a neutral solution OR to do something like "def/indef %1$s" where %1$s is the post type singular label.
This approach is used in wp core as well, e.g. https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2518 Or do you think this is a different case?

> I initially moved all the text to the abstract classes and injected the appropriate post type and singular or plural.

I think that was the right way to address this, less code, easier to maintain and dynamic.

pondermatic replied:
A big factor in the decision to abandon the work done on building text strings dynamically involved case folding. Take the [`llms_my_certificate` post type](https://github.com/gocodebox/lifterlms/blob/5.8.0/includes/class.llms.post-types.php#L790) for example. The names have each word capitalized, so we need to change the case depending on where it will be used in the sentence. For English, the first word in a sentence is usually capitalized. This rule is not used in all languages though. Worse yet, I found that in Greek, changing the case can change the spelling of a word. ["Awarded" = "Απονεμήθηκε", but "awarded" = "απονεμηθεί"](https://translate.google.com/?sl=en&tl=el&text=Awarded%20Certificate%0Aawarded%20certificate%0AAwarded%20certificate%0A&op=translate). Confusingly, changing the case of "Απονεμήθηκε Πιστοποιητικό" [seems to work](https://translate.google.com/?sl=el&tl=en&text=%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CE%A0%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%B1%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C%0A%CE%91%CF%80%CE%BF%CE%BD%CE%B5%CE%BC%CE%AE%CE%B8%CE%B7%CE%BA%CE%B5%20%CF%80%CE%B9%CF%83%CF%84%CE%BF%CF%80%CE%BF%CE%B9%CE%B7%CF%84%CE%B9%CE%BA%CF%8C&op=translate)?

Translating a bunch of similar text strings, while more work for the translators, allows for better translations.

eri-trabiccolo replied:
I understand your position. Sure your solution allows for better translations, but to me it's about balance.
How better? At what cost?
I gave you an example on how WordPress does basically the same thing I'm proposing in the rest api (an in other parts). 

You say:
> The names have each word capitalized, so we need to change the case depending on where it will be used in the sentence. For English, the first word in a sentence is usually capitalized.

Ok so the problem is that in our case we actually should "lowercase" the post type labels? Because something like
`'Sorry, you need to provide a valid Awarded Achievement ID.'` would be ugly?
Even if I understand the principle, we are talking about admin error messages that would show up like never or nearly never :D

And since we have to lowercase the post type labels, you found that...
> Worse yet, I found that in Greek, changing the case can change the spelling of a word. ["Awarded" = "Απονεμήθηκε", but "awarded" = "απονεμηθεί"](https://translate.google.com/?sl=en&tl=el&text=Awarded%20Certificate%0Aawarded%20certificate%0AAwarded%20certificate%0A&op=translate).

The real issue I find here is that we cannot use `strtolower()` with non utf8 charsets as the greek, but we should use [`mb_strtolower()`](https://www.php.net/manual/en/function.mb-strtolower.php) and I understand this is problematic as the `mb` suite is not built-in - but in this case we can just lowercase the string if the function exists and do nothing if it doesn't (rare cases).

What you say about changing the case in Greek can change the spelling of a word, is weird to me: you might be right, I don't know Greek and maybe you do, but certainly the link you provided doesn't prove your theory.
If you click on `απονεμηθεί πιστοποιητικό` translation you'll see that google suggests a variant:
`απονεμήθηκε πιστοποιητικό` were the first word is the lowercase version of `Απονεμήθηκε`.

I think that GoogleTranslate just tries to guess the "nature" of the first word, whether it's a name, an adjective (adverb) or a verb and for Google (but not necessarily for greek people who understand the context) the first capitalized word alters the "nature" of it.

You're basically saying that we cannot use the post type or taxonomy labels in localizable sentences (at least if we want to alter them with lowercase/uppercase), and that the mb_convert_case/mb_strtolower cannot be applied at least to the greek language.

Maybe you're right, I don't agree but it's fine. 
I don't think we have to spend more time on this: our opinion differ but there's nothing wrong with what you did so  I guess we can move on. :D

pondermatic replied:
@eri-trabiccolo wrote,
> I gave you an example on how WordPress does basically the same thing I'm proposing in the rest api (an in other parts).

I agree that injecting capitalized words in a REST API error message is not a problem. I'd be surprised if anyone translates those.

> Ok so the problem is that in our case we actually should "lowercase" the post type labels? Because something like
> `'Sorry, you need to provide a valid Awarded Achievement ID.'` would be ugly?
> Even if I understand the principle, we are talking about admin error messages that would show up like never or nearly never :D

I see your point. Having capitalized words in the middle of a sentence, while incorrect, is not that bad. In addition to admin error messages, this issue also affects [text in meta boxes](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php#L68). 

> The real issue I find here is that we cannot use strtolower() with non utf8 charsets as the greek, but we should use [mb_strtolower()](https://www.php.net/manual/en/function.mb-strtolower.php) and I understand this is problematic as the mb suite is not built-in - but in this case we can just lowercase the string if the function exists and do nothing if it doesn't (rare cases).

I wrote a class with functions to handle multibyte strings if the mb functions are available. They _usually_ are. If not, the built in string functions methods are used. I even wrote unit tests. I need to create a branch and push it up to the repo. 

> What you say about changing the case in Greek can change the spelling of a word, is weird to me: you might be right, I don't know Greek and maybe you do, but certainly the link you provided doesn't prove your theory.

The translator tools made me question whether changing the case could cause translation errors. I don't know the answer to that question, which is why, with a heavy heart,  I chose to abandon all the work I did on case conversion and building sentences in the abstract classes with concrete classes that return a full sentence.

> You're basically saying that we cannot use the post type or taxonomy labels in localizable sentences (at least if we want to alter them with lowercase/uppercase)

Yes, and I'm not very happy about it either.

eri-trabiccolo commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,309 @@
+		// There is no need to configure this meta box if we're not editing an engagement template or awarded engagement.
+		if (
+			is_null( $this->post ) ||
+			! in_array( $this->post->post_type, array( $this->post_type_awarded, $this->post_type_template ) )
```
```suggestion
			! in_array( $this->post->post_type, array( $this->post_type_awarded, $this->post_type_template ), true )
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,309 @@
+
+		if ( $this->post->post_type === $this->post_type_template ) {
+			$this->is_syncing_all_awarded_engagements = true;
+			$this->title = $this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENTS );
```
```suggestion
			$this->is_syncing_all_awarded_engagements = true;
			$this->title                                                          = $this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENTS );
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-meta-box-user-engagement-sync.php:
```diff
@@ -0,0 +1,309 @@
+			$this->title = $this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENTS );
+		} else {
+			$this->is_syncing_all_awarded_engagements = false;
+			$this->title = $this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENT );
```
```suggestion
			$this->is_syncing_all_awarded_engagements = false;
			$this->title                                                          = $this->get_text( self::TEXT_SYNC_AWARDED_ENGAGEMENT );
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-user-engagement.php:
```diff
@@ -93,12 +107,35 @@ public function get_earned_date( $format = null ) {
+			return false;
+		}
+
+		return new static( $post );
```
This is not semantically correct to me.
We're returning an user engagement object here: LLMS_User_Certificate/LLMS_User_Achievement models for llms_my_certificate/llms_my_achievement post types.
But I would expect a model for its template, so a model for llms_certificate/llms_achievement, which we actually don't have, maybe we should have them, maybe not and do otherwise, but this is not correct to me.
E.g. if I then call methods of the instance returned by this method I will end up checking on the wrong `$db_post_type` (see for instance https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/abstracts/abstract.llms.post.model.php#L636, this is just an example, we might even never use it)

pondermatic replied:
@eri-trabiccolo wrote
> We're returning an user engagement object here: LLMS_User_Certificate/LLMS_User_Achievement models for llms_my_certificate/llms_my_achievement post types.

You are correct. `LLMS_Abstract_User_Engagement::get_user_engagement_template()` returns `LLMS_User_Certificate`/`LLMS_User_Achievement` models. Although they are for template post types, not awarded. I agree that it is confusing and that `$db_post_type` should not be for the awarded engagement.

I will replace that method with [`LLMS_Trait_User_Engagement_Type::get_user_engagement()`](https://github.com/gocodebox/lifterlms/blob/1931-refactor-engagement-sync/includes/traits/llms-trait-user-engagement-type.php#L39) which can return either an awarded engagement or an engagement template. I used it in the other 3 types of classes.

There are 4 different types of classes:
1. controller
2. meta box
3. processor
4. model

Each has an abstract, an achievement and a certificate. It would be useful to have a class or trait for an engagement template and another class/trait for an awarded engagement.

I really need to diagram this all out to make certain we have the right methods and properties in the right classes/traits. It was easy to get confused during this refactoring and I fear there may be an architectural mess.

eri-trabiccolo commented on includes/abstracts/llms.abstract.privacy.php:
```diff
@@ -20,62 +20,63 @@
-	 * @version  3.18.0
+	 * @since 3.18.0
+	 *
+	 * @param string $name  plugin name
```
```suggestion
	 * @param string $name Plugin name.
	 * @return void
```

eri-trabiccolo commented on includes/abstracts/llms.abstract.privacy.php:
```diff
@@ -85,24 +86,26 @@ public function add_privacy_message() {
-	 * @return   false|LLMS_Student
-	 * @since    3.18.0
-	 * @version  3.18.0
+	 * @param string $email  Email address.
```
```suggestion
	 * @param string $email Email address.
```

eri-trabiccolo commented on includes/abstracts/llms.abstract.privacy.php:
```diff
@@ -112,52 +115,56 @@ protected static function get_student_by_email( $email ) {
-	 * @return   array
-	 * @since    3.18.0
-	 * @version  3.18.0
+	 * @param array $erasers  Existing erasers.
```
```suggestion
	 * @param array $erasers Existing erasers.
```

eri-trabiccolo commented on includes/abstracts/llms.abstract.privacy.php:
```diff
@@ -166,21 +173,21 @@ public function add_eraser( $id, $name, $callback ) {
-	 * @return   array
+	 * @param string $id       exporter ID
+	 * @param string $name     Human-readable exporter name
+	 * @param mixed  $callback Callback function (callable)
```
```suggestion
	 * @param string $id       Exporter ID.
	 * @param string $name     Human-readable exporter name.
	 * @param mixed  $callback Callback function (callable).
```

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php:
```diff
@@ -0,0 +1,121 @@
+	 * Type of user engagement.
+	 *
+	 * @since [version]
+	 *
```
our doc standards do not seem to want an empty line here.

pondermatic replied:
From the [LifterLMS documentation standards](https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md#tag-alignment-and-order):
> All changelog tags, `@since`, `@version`, and `@deprecated` should be grouped together with a space before the first `@since` tag and after the last tag in the group.

eri-trabiccolo replied:
yeah, you're right, as said before I was mislead by the example:
https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md#class-members
But you're right.

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php:
```diff
@@ -0,0 +1,121 @@
+
+	/**
+	 * Post types that this meta box should be added to.
+	 *
```
since?

pondermatic replied:
oops!

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php:
```diff
@@ -0,0 +1,121 @@
+	 * @param array $variables Optional variables that are used in sprintf().
+	 * @return string
+	 */
+	protected function get_text( $text_type, $variables = array() ) {
```
Similarly to what I said for the controllers' get_text, I wonder if we need this method here and in the certificate sync metabox. Aren't they identical if not for the `achievement|certificate` post type label (s)?

pondermatic replied:
Injecting a translated post type label into a translated sentence is bad.

From the [WordPress internationalization handbook](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#best-practices-for-writing-strings):
> Use entire sentences – in most languages word order is different than that in English.

eri-trabiccolo replied:
I don't think this is the case to be honest, see my comment here:
https://github.com/gocodebox/lifterlms/pull/1946#discussion_r803580875

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/class-llms-meta-box-certificate-sync.php:
```diff
@@ -0,0 +1,121 @@
+	 * Type of user engagement.
+	 *
+	 * @since [version]
+	 *
```
```suggestion
```
and so on

eri-trabiccolo commented on includes/admin/reporting/class.llms.admin.reporting.php:
```diff
@@ -329,54 +330,45 @@ private function get_tab_cap( $tab = null ) {
-	 * @return   void
-	 * @since    3.2.0
-	 * @version  3.16.0
+	 * @since 3.2.0
+	 * @since [version] Removed loading of class files that don't instantiate their class in favor of autoloading.
```
```suggestion
	 * @since 3.2.0
	 * @since 3.16.0 Unknown.
	 * @since [version] Removed loading of class files that don't instantiate their class in favor of autoloading.
```

pondermatic replied:
Thank you @eri-trabiccolo! I knew I accidently deleted a changelog entry somewhere, but I couldn't find it later on.

eri-trabiccolo commented on includes/admin/reporting/class.llms.admin.reporting.php:
```diff
@@ -329,54 +330,45 @@ private function get_tab_cap( $tab = null ) {
-	 * @since    3.2.0
-	 * @version  3.19.4
+	 * @since 3.2.0
+	 *
```
```suggestion
	 * @since 3.2.0
	 * @since 3.19.4 Unknown.
	 *
```

eri-trabiccolo commented on includes/admin/reporting/widgets/class.llms.analytics.widget.ajax.php:
```diff
@@ -24,6 +24,7 @@ class LLMS_Analytics_Widget_Ajax {
 	 * @since 3.0.0
```
This file's header version tag should be udpated.

eri-trabiccolo commented on includes/class-llms-loader.php:
```diff
@@ -253,30 +311,20 @@ public function includes() {
 	 * @since 4.12.0 Class `LLMS_Staging` always loaded instead of only loaded on admin panel.
 	 * @since 5.0.0 Include `LLMS_Forms_Unsupported_Versions` class.
 	 * @since [version] Drop usage of deprecated `FILTER_SANITIZE_STRING`.
+	 *              Include `LLMS_Abstract_Meta_Box_User_Engagement_Sync`.
```
```suggestion
```

eri-trabiccolo commented on includes/processors/class-llms-processor-certificate-sync.php:
```diff
@@ -0,0 +1,71 @@
+	 * @param array $variables Optional variables that are used in sprintf().
+	 * @return string
+	 */
+	protected function get_text( $text_type, $variables = array() ) {
```
completely equal to LLMS_Processor_Achievement_Sync::get_text() except for the post type label?

eri-trabiccolo commented on includes/abstracts/abstract.llms.admin.metabox.php:
```diff
@@ -10,11 +10,6 @@
 
 defined( 'ABSPATH' ) || exit;
 
-// Include all classes for each of the metabox types.
```
Mmm,
I edited a course and got all the llms blocks broken saying: Your site doesn't include support for the "llms/pricing-table" block...
so I checked the error log and there was this:
```
[09-Feb-2022 13:22:18 UTC] PHP Fatal error:  Uncaught Error: Class "LLMS_Metabox_Post_Excerpt_Field" not found in /Users/rock/Dev/plugins/lifterlms/includes/abstracts/abstract.llms.admin.metabox.php:388
Stack trace:
#0 /Users/rock/Dev/plugins/lifterlms/includes/abstracts/abstract.llms.admin.metabox.php(354): LLMS_Admin_Metabox->process_field(Array)
#1 /Users/rock/Dev/plugins/lifterlms/includes/abstracts/abstract.llms.admin.metabox.php(269): LLMS_Admin_Metabox->process_fields()
#2 /shared/httpd/mywp/wordpress.git/wp-admin/includes/template.php(1395): LLMS_Admin_Metabox->output(Object(WP_Post), Array)
#3 /shared/httpd/mywp/wordpress.git/wp-admin/includes/post.php(2296): do_meta_boxes(Object(WP_Screen), 'normal', Object(WP_Post))
#4 /shared/httpd/mywp/wordpress.git/wp-admin/edit-form-blocks.php(310): the_block_editor_meta_boxes()
#5 /shared/httpd/mywp/wordpress.git/wp-admin/post.php(187): require('/shared/httpd/m...')
#6 {main}
  thrown in /Users/rock/Dev/plugins/lifterlms/includes/abstracts/abstract.llms.admin.metabox.php on line 388
```

can you reproduce it?

pondermatic replied:
@eri-trabiccolo, thank you for reporting this issue. I am able to reproduce it and have fixed it.

eri-trabiccolo commented on includes/processors/class-llms-processor-achievement-sync.php:
```diff
@@ -0,0 +1,71 @@
+				return sprintf(
+					/* translators: %1$s: opening anchor tag that links to the achievement template, %2$s: achievement template name, %3$d: achievement template ID, %4$s: closing anchor tag */
+					__( 'Awarded Achievements sync completed for the template %1$s%2$s (#%3$d)%4$s.', 'lifterlms' ),
+					'<a href="' . get_edit_post_link( ( $template_id ) ) . '" target="_blank">',
```
Some weeks ago (I talked with Thomas about this but I didn't open an issue) I've found that the `get_edit_post_link()` here returns an empty string b/c the completion happens in the background, no user is logged in and the edit link is not set since we don't show the ui to users who cannot manage the awards. I think we should build our simple/rough version of this function for these post types.

pondermatic replied:
I agree.

pondermatic replied:
Another option is to use the `user_has_cap` filter hook in `WP_User::has_cap()`. We could register a filter right before we call `get_edit_post_link()` in `LLMS_Processor_Certificate_Sync::get_text()` and then remove the filter.

eri-trabiccolo replied:
I'm not totally sure this will work. Because actually `get_edit_post_link()` relies on the post object property `_edit_link` value, which is empty (or rather not set) when the post type is registered and `show_ui` is false. But maybe I'm wrong.

pondermatic replied:
I thought `WP_Post_Type::$_edit_link` would be set. But you are right, it is not. It gets set to an empty string [here](https://github.com/WordPress/wordpress-develop/blob/5.9/src/wp-includes/class-wp-post-type.php#L524) because `show_ui` is set based on the result of `current_user_can()`.

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/class-llms-meta-box-achievement-sync.php:
```diff
@@ -0,0 +1,121 @@
+				return sprintf(
+					/* translators: %1$s: link to edit the achievement template, %2$s: closing anchor tag */
+					__( 'Sync this awarded achievement with its %1$sachievement template%2$s.', 'lifterlms' ),
+					'<a href="' . get_edit_post_link( ( $variables['template_id'] ?? 0 ) ) . '" target="_blank">',
```
I've tested this, and the link here doesn't point to the template but to the awarded post itself.
The same happens with the certificate.

eri-trabiccolo reviewed (changes requested):
Aside from the changes requested (and the ones requested by Thomas) the sync works good to me.
You did a massive work here! 
We need to test the files loading changes a little bit more I think (see the issue I reported [here](https://github.com/gocodebox/lifterlms/pull/1946#discussion_r802661378)).

pondermatic committed:
Give the template ID to get_text() for sync description on this awarded engagement.

pondermatic committed:
Revert commit e23ebf588035fac04e414f3cf35807fc1411fcfc.

pondermatic committed:
Add DocBlock summaries to new constants.

pondermatic committed:
Add '_TITLE_' to the title constants.

pondermatic committed:
Remove outdated changelog entry.

pondermatic committed:
Do not use backslash in localized string.

pondermatic committed:
Fix loading meta box field classes.

pondermatic committed:
Search array strictly.

pondermatic committed:
Rename is_syncing_all_awarded_engagements property to is_current_post_a_template.

pondermatic committed:
Cleanup DocBlocks.

pondermatic committed:
Add LLMS_Abstract_Processor::get_edit_post_link().

pondermatic committed:
Fix edit template link.

pondermatic committed:
Move notices about scheduling to LLMS_Abstract_Processor_User_Engagement_Sync::get_text().

pondermatic committed:
Move notices about scheduling to LLMS_Abstract_Processor_User_Engagement_Sync::get_text().

pondermatic committed:
Replace LLMS_Abstract_User_Engagement::get_user_engagement_template() with LLMS_Trait_User_Engagement_Type::get_user_engagement().
